### ISSUE TITLE
Support `... | ""` as alias for `Opt(...)`

### ISSUE THREAD

**Issue by anntzer** [pyparsing/pyparsing#412]
Depending on the context, `<something> | ""` (or `"" | <something>`) may be a clearer way to spell out `Opt(...)`; I believe this *almost* works except for the fact that `Literal("")` currently raises an error asking to use `Empty()` instead.  Could this restriction be lifted?

I see various ways to lift the restruction, depending on what you are exactly willing to accept.  One would be to just directly allow `Literal("")` (if you worry about the loss in perf, you *could* define `Literal.__new__` to return `Empty()` in that case, although that's a bit too magical in my taste).  Or you could change `ParseElementEnhance.__init__` at https://github.com/pyparsing/pyparsing/blob/966d6fded149c6c11993746b0d72166bc04e4504/pyparsing/core.py#L4345-L4353 to special-case `expr == ""` there (or rather `len(expr) == 0` to support bytes too) and return `Empty()`.  Finally, you could also change `ParserElement.__or__/__ror__` to detect that case and construct an `Optional(...)` as needed (this may be the simplest from an explanation point of view: "`|` constructs a MatchFirst, except that `| ""` constructs an Optional").

I can make a PR if you agree with any of these solutions.

**Comment by ptmcg** [comment]
I'll look these ideas over, but it won't be for a few days.

Could you put together some examples showing how `expr | ""` or `expr | Empty()` is preferable to `Opt(expr)`?  (Granted, this seems to be paralleling the discussion around `typing.Optional[SomeType]` being better written as `SomeType | None`.) 

Doesn't `expr | Empty()` work with the current code, without making any changes to Literal?

I guess `Empty()` is not quite so easy to write as `""`, and requires an import, or writing as `pp.Empty()`. What about `expr | None`? And have the `__or__` operator recognize a `None` operand and replace it with `Empty()`?  (Still need to give this a little more thought, but this seems like it might address your use case.)

**Comment by anntzer** [comment]
For example, consider matplotlib's tex-parser element for tex's genfrac command (`\genfrac{left_delim}{right_delim}{rulesize}{style}{num}{denum}`) -- not a full example, other `p.foo` elements are defined earlier:
```python
        p.genfrac <<= r"\genfrac" - (
            "{" + Optional(p.ambi_delim | p.left_delim)("ldelim") + "}"
            + "{" + Optional(p.ambi_delim | p.right_delim)("rdelim") + "}"
            + "{" + p.float_literal("rulesize") + "}"
            + p.optional_group("style")
            + p.required_group("num")
            + p.required_group("den") | Error(...))
```
I would argue that it would read better as
```python
        p.genfrac <<= r"\genfrac" - (
            "{" + (p.ambi_delim | p.left_delim | "")("ldelim") + "}"
            + "{" + (p.ambi_delim | p.right_delim | "")("rdelim") + "}"
            + "{" + p.float_literal("rulesize") + "}"
            + p.optional_group("style")
            + p.required_group("num")
            + p.required_group("den") | Error(...))
```
(an empty delimiter is just another possible delimiter, that case can be mentioned last rather than "first") although this may certainly depend on taste.

`expr | Empty()` requires an additional import and is longer to write.  `expr | None` could perhaps also work indeed; however, one important point if favor of `expr | ""`, in my view, is that (unless I am mistaken) this seems currently only made impossible by the specific check `if <match_string is empty>: raise ValueError("null string passed to Literal; use Empty() instead")`; if you removed that check I don't think(?) *any* of the semantics of pyparsing would be modified and none of the docs need to be changed (unless you introduce optimizations that replace the empty Literal by an Empty or an Optional, but this is somewhat transparent to most end users).

**Comment by anntzer** [comment]
Closed by #413, AFAICT.  Thanks!
